### ISSUE TITLE
Differentiate restart from prev

### ISSUE THREAD

**Issue by ericmarkmartin** [hnarayanan/shpotify#29]
I have found the fact that previous track actually restarts the song if you are a substantial amount through it to be an annoying feature amongst many audio control solutions, and while I often believe it to be a necessary evil (Mac laptops have only three audio buttons, the same with in-line headset controls), with a command line interface one should have the luxury of separate functions.

**Comment by hnarayanan** [comment]
Yes, this is super annoying.

Do you know of a proper signal to send Spotify (the app) to inform it not to restart the song but to jump to the previous track?

**Comment by IndicaInkwell** [comment]
Calling previous track 2 times after 3 or 4 seconds into the song seems to work. Also calling previous track repeatedly doesn't seem to make the transition jittery or anything.
```applescript
tell application "Spotify"
	previous track
	if player position is greater than 3 then
		previous track
	end if
end tell
```

**Comment by ericmarkmartin** [comment]
@IndicaInkwell, that looks like a good solution to me. However, shouldn't the if statement go first, as `previous track` will take the user to the beginning of a track (be it their current track or the previous track), and thus the if will always evaluate.

In terms of a restart function, I was thinking:

    tell application "Spotify"
        if player position is less than 3 then
            next track
        end if
    end tell

Of course, this still faces an issue in that the `next track` command counts as a skip towards your next ad AFAIK.

On another note, is there any way we can get an exact value instead of 3?

As a final comment, I don't think this is a bug, as this has to do with Spotify's (and nearly every other music management software's) design, not the script's.

**Comment by IndicaInkwell** [comment]
@ericmarkmartin you're totally right. It worked for me because the properties have a big delay before being updated to the new values.

A restart function could also be implemented like
```applescript
tell application "Spotify"
	play track (get spotify url of current track)
end tell
```

I'll try some tests and see if I can get a more precise number than 3 or 4.

**Comment by ericmarkmartin** [comment]
@IndicaInkwell Sounds great. Nice thinking with the restart function.

**Comment by hnarayanan** [comment]
(I am really excited that you guys are iterating and figuring this out.)

**Comment by ericmarkmartin** [comment]
I realized it's been almost a year since anyone touched this (which I also realized was quite ironic considering @hnarayanan's last comment) so I cooked up #54 to fix this issue.

**Comment by IndicaInkwell** [comment]
Oh man, it has almost been a year! I apologize for completely neglecting this and to be honest I also completely forgot about it.

**Comment by ericmarkmartin** [comment]
No worries, I had as well until I looked at the open PR’s. As of now, I used "double previous" and your "play the current track” methods for `spotify prev` and `spotify replay` respectively, but if you can think of a better way (the current way works fine but I believe counts towards the next ad count for free users), please let me know and we’ll update the code.

**Comment by IndicaInkwell** [comment]
That sounds good, will keep that in mind.

I did notice something right now though: for replaying the track do you know if, and why, it may be better to `play track (get spotify url of current track)` or to `set player position to 0`? I can't think of why, but I feel like there is a reason.

Setting player position doesn't increment ad count afaik.

**Comment by hnarayanan** [comment]
I like where @IndicaInkwell is going with this. Feel free to reopen this issue and update the code if you guys come to an agreement.

**Comment by IndicaInkwell** [comment]
@hnarayanan Sweet, will do!

For going to the previous track you could do something like this:

```
tell application "Spotify"
	set player position to 0
	previous track
end tell
```

Which will remove the guessing of the correct time frame to go back a song?

**Comment by hnarayanan** [comment]
I like the sound of this too, if it works without any obvious skipping in the audio.

**Comment by IndicaInkwell** [comment]
@hnarayanan I'm gonna fork and play/use it for a day or two and get back to you. But, I guess as the saying goes, why fix what aint broke?

**Comment by hnarayanan** [comment]
There is no hurry and no pressure. Play around and we'll merge if/when you arrive at something better.

**Comment by IndicaInkwell** [comment]
All right, I have been using shpotify with the above changes and it works just fine. It also seems like neither `set player position to 0` nor `play track (get spotify url of current track)` goes towards the next ad.

An advantage of setting player pos is that it's immediate. The play track version takes a moment.

What do guys think? Shall I open a pull request?

**Comment by hnarayanan** [comment]
I don't have a strong feeling about this because I tend to listen to whatever same song I'm listening to over and over again on repeat. Meaning I don't manually use the replay option. 🤓

So I'll leave someone else to comment more meaningfully.

**Comment by ericmarkmartin** [comment]
I think I agree with @IndicaInkwell on this one. Setting player position in addition to avoiding the unconfirmed 3 seconds value semantically seems more directly like what we are after.

**Comment by IndicaInkwell** [comment]
Okay, cool. I'm going to open a pull request in a little bit.

**Comment by davidem** [comment]
For future reference: this bug has been fixed with PR #54 and can be closed.
